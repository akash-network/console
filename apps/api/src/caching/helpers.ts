import * as Sentry from "@sentry/node";
import { differenceInSeconds } from "date-fns";

import { LoggerService } from "@src/core";
import MemoryCacheEngine from "./memoryCacheEngine";

const logger = new LoggerService({ context: "Caching" });

export const cacheEngine = new MemoryCacheEngine();
const pendingRequests: { [key: string]: Promise<unknown> } = {};

interface CachedObject<T> {
  date: Date;
  data: T;
}

interface MemoizeOptions {
  keepData?: boolean;
  ttlInSeconds?: number;
  key?: string;
}

export const Memoize = (options?: MemoizeOptions) => (target: object, propertyName: string, descriptor: PropertyDescriptor) => {
  const originalMethod = descriptor.value;

  const cacheKey = options?.key || `${target.constructor.name}#${propertyName}`;

  descriptor.value = async function memoizedFunction(...args: unknown[]) {
    return await cacheResponse(options?.ttlInSeconds || 60 * 2, cacheKey, originalMethod.bind(this, ...args), options?.keepData);
  };
};

export async function cacheResponse<T>(seconds: number, key: string, refreshRequest: () => Promise<T>, keepData?: boolean): Promise<T> {
  const duration = seconds * 1000;
  const cachedObject = cacheEngine.getFromCache(key) as CachedObject<T> | undefined;
  logger.debug(`Request for key: ${key}`);

  // If first time or expired, must refresh data if not already refreshing
  const cacheExpired = Math.abs(differenceInSeconds(cachedObject?.date, new Date())) > seconds;
  if ((!cachedObject || cacheExpired) && !(key in pendingRequests)) {
    logger.debug(`Object was not in cache or is expired, making new request for key: ${key}`);
    pendingRequests[key] = refreshRequest()
      .then(data => {
        cacheEngine.storeInCache(key, { date: new Date(), data: data }, keepData ? undefined : duration);
        return data;
      })
      .catch(err => {
        if (cachedObject) {
          logger.error(`Error making cache request ${err}`);
          Sentry.captureException(err);
        } else {
          throw err;
        }
      })
      .finally(() => {
        delete pendingRequests[key];
      });
  }

  // If there is data in cache, return it even if it is expired. Otherwise, wait for the refresh request to finish
  if (cachedObject) {
    logger.debug(`Returning cached object for key: ${key}`);
    return cachedObject.data;
  } else {
    logger.debug(`Waiting for pending request for key: ${key}`);
    return (await pendingRequests[key]) as T;
  }
}

export const cacheKeys = {
  getProviderGraphData: "getProviderGraphData",
  web3IndexRevenue: "web3IndexRevenue",
  getProviderActiveLeasesGraphData: "getProviderActiveLeasesGraphData",
  getProviderAttributesSchema: "getProviderAttributesSchema",
  getTemplates: "getTemplates",
  getMarketData: "getMarketData",
  getAuditors: "getAuditors",
  getProviderList: "getProviderList",
  getChainStats: "getChainStats",
  getProviderRegions: "getProviderRegions",
  getMainnetNodes: "getMainnetNodes",
  getTestnetNodes: "getTestnetNodes",
  getSandboxNodes: "getSandboxNodes",
  getMainnetVersion: "getMainnetVersion",
  getTestnetVersion: "getTestnetVersion",
  getSandboxVersion: "getSandboxVersion",
  getGpuModels: "getGpuModels"
};
