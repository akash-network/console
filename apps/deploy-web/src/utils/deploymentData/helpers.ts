import { SDL } from "@akashnetwork/akashjs/build/sdl";
import type { v2Sdl } from "@akashnetwork/akashjs/build/sdl/types";
import type { NetworkId } from "@akashnetwork/akashjs/build/types/network";
import axios from "axios";

export class CustomValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "CustomValidationError";
  }
}

const specSuffixes = {
  Ki: 1024,
  Mi: 1024 * 1024,
  Gi: 1024 * 1024 * 1024,
  Ti: 1024 * 1024 * 1024 * 1024,
  Pi: 1024 * 1024 * 1024 * 1024 * 1024,
  Ei: 1024 * 1024 * 1024 * 1024 * 1024 * 1024,
  K: 1000,
  M: 1000 * 1000,
  G: 1000 * 1000 * 1000,
  T: 1000 * 1000 * 1000 * 1000,
  P: 1000 * 1000 * 1000 * 1000 * 1000,
  E: 1000 * 1000 * 1000 * 1000 * 1000 * 1000,
  Kb: 1000,
  Mb: 1000 * 1000,
  Gb: 1000 * 1000 * 1000,
  Tb: 1000 * 1000 * 1000 * 1000,
  Pb: 1000 * 1000 * 1000 * 1000 * 1000,
  Eb: 1000 * 1000 * 1000 * 1000 * 1000 * 1000
};

export async function getCurrentHeight(apiEndpoint: string) {
  const response = await axios.get(`${apiEndpoint}/blocks/latest`);
  const data = response.data;

  const height = parseInt(data.block.header.height);
  return height;
}

export function parseSizeStr(str: string) {
  try {
    const suffix = Object.keys(specSuffixes).find(s => str.toLowerCase().endsWith(s.toLowerCase()));

    if (suffix) {
      const suffixPos = str.length - suffix.length;
      const numberStr = str.substring(0, suffixPos);
      return (parseFloat(numberStr) * specSuffixes[suffix as keyof typeof specSuffixes]).toString();
    } else {
      return parseFloat(str);
    }
  } catch (err) {
    console.error(err);
    throw new Error("Error while parsing size: " + str);
  }
}

type NetworkType = "beta2" | "beta3";

function isValidString(value: unknown): value is string {
  return typeof value === "string" && !!value;
}

export function getSdl(yamlJson: string | v2Sdl, networkType: NetworkType, networkId: NetworkId) {
  return isValidString(yamlJson) ? SDL.fromString(yamlJson, networkType, networkId) : new SDL(yamlJson, networkType, networkId);
}

export function DeploymentGroups(yamlJson: string | v2Sdl, networkType: NetworkType, networkId: NetworkId) {
  const sdl = getSdl(yamlJson, networkType, networkId);
  return sdl.groups();
}

export function Manifest(yamlJson: string | v2Sdl, networkType: NetworkType, networkId: NetworkId, asString = false) {
  const sdl = getSdl(yamlJson, networkType, networkId);
  return sdl.manifest(asString);
}

export async function ManifestVersion(yamlJson: string | v2Sdl, networkType: NetworkType, networkId: NetworkId) {
  const sdl = getSdl(yamlJson, networkType, networkId);
  return sdl.manifestVersion();
}

export function ManifestYaml(sdlConfig: v2Sdl, networkType: NetworkType, networkId: NetworkId) {
  const sdl = getSdl(sdlConfig, networkType, networkId);
  return sdl.manifestSortedJSON();
}
