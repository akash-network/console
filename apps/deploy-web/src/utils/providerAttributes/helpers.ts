import { nanoid } from "nanoid";
import type { z } from "zod";

import type {
  ProviderAttributeSchemaDetail,
  ProviderAttributeSchemaDetailValue,
  providerAttributesFormValuesSchema,
  ProviderAttributesSchema
} from "@src/types/providerAttributes";

/**
 * Maps the form values to the attributes that are broadcasted to the network
 * @param data
 * @param providerAttributesSchema
 * @returns
 */
export const mapFormValuesToAttributes = (data: z.infer<typeof providerAttributesFormValuesSchema>, providerAttributesSchema: ProviderAttributesSchema) => {
  const attributes: { key: string; value: string }[] = [];

  Object.keys(data).forEach(key => {
    const value = data[key as keyof typeof data];
    const attribute = providerAttributesSchema[key as keyof ProviderAttributesSchema] as ProviderAttributeSchemaDetail;
    if (attribute && value) {
      switch (attribute.type) {
        case "string":
        case "number":
        case "boolean":
          attributes.push({ key, value: `${value}` });
          break;
        case "option":
          // eslint-disable-next-line no-case-declarations
          const detailValue = value as unknown as ProviderAttributeSchemaDetailValue;
          // eslint-disable-next-line no-case-declarations
          const attributeValue = attribute.values?.find(v => v.key === detailValue.key);
          attributes.push({ key: attribute.key, value: `${attributeValue?.key}` });
          break;
        case "multiple-option":
          // eslint-disable-next-line no-case-declarations
          const values = value as ProviderAttributeSchemaDetailValue[];
          values.forEach(_val => {
            const attributeValue = attribute.values?.find(v => v.key === _val.key);

            if (attributeValue?.key) {
              attributes.push({ key: attributeValue.key, value: `${attributeValue?.value}` });
            }
          });
          break;
        default:
          break;
      }
    }

    if (key === "unknown-attributes") {
      const unknownAttributes = value as { id: string; key: string; value: string }[];
      unknownAttributes.forEach(x => attributes.push({ key: x.key, value: x.value }));
    }
  });

  return attributes;
};

/**
 * Get the list of attributes that are unknown to the schema
 * @param attributes
 * @param providerAttributesSchema
 * @returns
 */
export const getUnknownAttributes = (attributes: { key: string; value: string }[], providerAttributesSchema: ProviderAttributesSchema) => {
  const possibleAttributes = Object.values(providerAttributesSchema)
    .map(x => {
      switch (x.type) {
        case "string":
        case "number":
        case "boolean":
        case "option":
          return x.key;
        case "multiple-option":
          return x.values?.map(v => v.key);
        default:
          return null;
      }
    })
    .filter(x => x)
    .flat();

  const res = attributes.filter(x => !possibleAttributes.includes(x.key)).map(x => ({ id: nanoid(), key: x.key, value: x.value }));

  return res;
};
