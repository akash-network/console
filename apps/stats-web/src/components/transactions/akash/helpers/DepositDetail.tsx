import { type Deposit } from "@akashnetwork/chain-sdk/private-types/akash.v1";
import { Source as DepositSource } from "@akashnetwork/chain-sdk/private-types/akash.v1";
import type { Coin } from "@akashnetwork/chain-sdk/private-types/cosmos.v1beta1";

import { AKTAmount } from "@/components/AKTAmount";
import { LabelValue } from "@/components/LabelValue";

const isCoin = (value: unknown): value is Coin => !value || (typeof value === "object" && "denom" in value && "amount" in value);
export const DepositDetail: React.FunctionComponent<{ value: Coin | Deposit }> = ({ value }) => {
  if (isCoin(value)) {
    return <LabelValue label="Deposit" value={<AKTAmount uakt={value.amount} showAKTLabel showUSD />} />;
  }

  return (
    <>
      <LabelValue label="Deposit Amount" value={<AKTAmount uakt={value.amount.amount} showAKTLabel showUSD />} />
      <LabelValue label="Deposit Sources" value={value.sources.map(source => DepositSource[source]).join(", ")} />
    </>
  );
};
