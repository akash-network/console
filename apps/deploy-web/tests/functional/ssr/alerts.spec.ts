import { AppHttpClient } from "@tests/helpers/AppHttpClient";

describe("Alerts", () => {
  describe("GET /alerts", () => {
    it("display alerts for logged in user", async () => {
      const client = await setup({ authenticated: true, withManagedWallet: true });

      const response = await client.get(`/alerts`);

      expect(response.status).toBe(200);
      expect(response.data).toMatchSnapshot();
    });

    it("displays 404 for non authenticated user", async () => {
      const client = await setup({ authenticated: false, withManagedWallet: true });

      const response = await client.get("/alerts");

      expect(response.status).toBe(404);
      expect(response.data).toMatchSnapshot();
    });
  });

  async function setup(input?: { authenticated?: boolean; withManagedWallet?: boolean }) {
    const client = new AppHttpClient("http://localhost:3000");

    if (input?.authenticated) {
      await client.loginViaOauth();
    }

    if (input?.withManagedWallet) {
      await client.startTrial();
    }

    return client;
  }
});
