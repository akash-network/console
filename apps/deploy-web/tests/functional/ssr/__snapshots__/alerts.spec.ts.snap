// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alerts GET /alerts display alerts for logged in user 1`] = `"<!DOCTYPE html><html lang="en-US"><head><style data-next-hide-fouc="true">body{display:none}</style><noscript data-next-hide-fouc="true"><style>body{display:block}</style></noscript><meta charSet="utf-8"/><meta name="viewport" content="minimum-scale=1, initial-scale=1, width=device-width, shrink-to-fit=no"/><link rel="icon" href="/favicon.ico"/><title>Akash Console</title><meta name="description" content="Akash Console is the #1 platform to deploy docker containers on the Akash Network, a decentralized super cloud compute marketplace. Explore, deploy and track all in one place!"/><meta name="twitter:card" content="summary_large_image"/><meta name="twitter:site" content="@akashnet_"/><meta name="twitter:creator" content="@akashnet_"/><meta property="og:title" content="Akash Console"/><meta property="og:description" content="Deploy docker containers on the decentralized supercloud Akash Network."/><meta property="og:url" content="https://console.akash.network/"/><meta property="og:type" content="website"/><meta property="og:image" content="https://console.akash.network/akash-console.png"/><meta property="og:image:alt" content="AkashConsole Cover Image"/><meta property="og:image:width" content="1200"/><meta property="og:image:height" content="630"/><meta property="og:locale" content="en_US"/><meta property="og:site_name" content="Akash Console"/><meta name="robots" content="index,follow"/><meta name="next-head-count" content="19"/><meta name="emotion-insertion-point" content=""/><style data-emotion="mui "></style><link rel="shortcut icon" href="/favicon.ico"/><link rel="manifest" href="/manifest.json" crossorigin="use-credentials"/><link rel="mask-icon" href="/safari-pinned-tab.svg" color="#0c0b0b"/><meta name="msapplication-TileColor" content="#0c0b0b"/><link rel="apple-touch-icon" href="/apple-touch-icon.png"/><meta name="theme-color" content="#0c0b0b"/><link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png"/><link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png"/><link href="https://fonts.googleapis.com/css2?family=Inter:wght@100;300;500;700;900&amp;display=swap" rel="stylesheet"/><link rel="preload" href="/_next/static/media/e11418ac562b8ac1-s.p.woff2" as="font" type="font/woff2" crossorigin="anonymous" data-next-font="size-adjust"/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills.js"></script><script src="/_next/static/chunks/webpack.js" defer=""></script><script src="/_next/static/chunks/main.js" defer=""></script><script src="/_next/static/chunks/pages/_app.js" defer=""></script><script src="/_next/static/chunks/pages/alerts.js" defer=""></script><script src="/_next/static/development/_buildManifest.js" defer=""></script><script src="/_next/static/development/_ssgManifest.js" defer=""></script><noscript id="__next_css__DO_NOT_USE__"></noscript></head><body><div id="__next"><main class="h-full bg-background font-sans tracking-wide antialiased __variable_3a0388"><script>!function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&true)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}if(e==='light'||e==='dark')d.style.colorScheme=e}catch(e){}}()</script><div style="visibility:hidden"></div></main></div><script src="/_next/static/chunks/react-refresh.js"></script><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{},"__N_SSP":true},"page":"/alerts","query":{},"buildId":"development","isFallback":false,"gssp":true,"locale":"en-US","locales":["en-US"],"defaultLocale":"en-US","scriptLoader":[]}</script></body></html>"`;

exports[`Alerts GET /alerts displays 404 for non authenticated user 1`] = `"<!DOCTYPE html><html lang="en-US"><head><style data-next-hide-fouc="true">body{display:none}</style><noscript data-next-hide-fouc="true"><style>body{display:block}</style></noscript><meta charSet="utf-8"/><meta name="viewport" content="minimum-scale=1, initial-scale=1, width=device-width, shrink-to-fit=no"/><link rel="icon" href="/favicon.ico"/><title>Akash Console</title><meta name="description" content="Akash Console is the #1 platform to deploy docker containers on the Akash Network, a decentralized super cloud compute marketplace. Explore, deploy and track all in one place!"/><meta name="twitter:card" content="summary_large_image"/><meta name="twitter:site" content="@akashnet_"/><meta name="twitter:creator" content="@akashnet_"/><meta property="og:title" content="Akash Console"/><meta property="og:description" content="Deploy docker containers on the decentralized supercloud Akash Network."/><meta property="og:url" content="https://console.akash.network/"/><meta property="og:type" content="website"/><meta property="og:image" content="https://console.akash.network/akash-console.png"/><meta property="og:image:alt" content="AkashConsole Cover Image"/><meta property="og:image:width" content="1200"/><meta property="og:image:height" content="630"/><meta property="og:locale" content="en_US"/><meta property="og:site_name" content="Akash Console"/><meta name="robots" content="index,follow"/><meta name="next-head-count" content="19"/><meta name="emotion-insertion-point" content=""/><style data-emotion="mui "></style><link rel="shortcut icon" href="/favicon.ico"/><link rel="manifest" href="/manifest.json" crossorigin="use-credentials"/><link rel="mask-icon" href="/safari-pinned-tab.svg" color="#0c0b0b"/><meta name="msapplication-TileColor" content="#0c0b0b"/><link rel="apple-touch-icon" href="/apple-touch-icon.png"/><meta name="theme-color" content="#0c0b0b"/><link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png"/><link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png"/><link href="https://fonts.googleapis.com/css2?family=Inter:wght@100;300;500;700;900&amp;display=swap" rel="stylesheet"/><link rel="preload" href="/_next/static/media/e11418ac562b8ac1-s.p.woff2" as="font" type="font/woff2" crossorigin="anonymous" data-next-font="size-adjust"/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills.js"></script><script src="/_next/static/chunks/webpack.js" defer=""></script><script src="/_next/static/chunks/main.js" defer=""></script><script src="/_next/static/chunks/pages/_app.js" defer=""></script><script src="/_next/static/chunks/pages/404.js" defer=""></script><script src="/_next/static/development/_buildManifest.js" defer=""></script><script src="/_next/static/development/_ssgManifest.js" defer=""></script><noscript id="__next_css__DO_NOT_USE__"></noscript></head><body><div id="__next"><main class="h-full bg-background font-sans tracking-wide antialiased __variable_3a0388"><script>!function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&true)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}if(e==='light'||e==='dark')d.style.colorScheme=e}catch(e){}}()</script><div style="visibility:hidden"></div></main></div><script src="/_next/static/chunks/react-refresh.js"></script><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{}},"page":"/404","query":{},"buildId":"development","nextExport":true,"autoExport":true,"isFallback":false,"locale":"en-US","locales":["en-US"],"defaultLocale":"en-US","notFoundSrcPage":"/alerts","scriptLoader":[]}</script></body></html>"`;
